packages/loopring_v3/contracts/amm/libamm/AmmJoinProcess.sol

```diff
@@ -83,109 +37,178 @@ library AmmJoinProcess
-                    require(transfer.storageID == join.storageIDs[i], "INVALID_TX_DATA");
-                }
-
-                // Now approve this transfer
                 transfer.validUntil = 0xffffffff;
                 bytes32 txHash = TransferTransaction.hashTx(ctx.exchangeDomainSeparator, transfer);
                 ctx.exchange.approveTransaction(join.owner, txHash);
-
-                ctx.ammActualL2Balances[i] = ctx.ammActualL2Balances[i].add(amount);
-
             } else {
-                // Make the amount unavailable for withdrawing
-                address token = ctx.tokens[i].addr;
-                S.userBalance[token][join.owner] = S.userBalance[token][join.owner].sub(amount);
+                for (uint j = 0; j < ctx.size; j++) {
+                    address token =  ctx.tokens[j].addr;
+                    S.addUserBalance(ctx.exchange.owner(), token, join.joinFees[j]);
+                    S.addUserBalance(join.owner, token, join.joinAmounts[j].sub(amounts[j]));
+                }
             }
+        }
 
-            ctx.ammExpectedL2Balances[i] = ctx.ammExpectedL2Balances[i].add(amount);
+        // Handle pool tokens
+        if (join.mintToLayer2) {
+            S.mint(address(this), mintAmount);
+            _approvePoolTokenDeposit(ctx, mintAmount, join.owner);
+        } else {
+            S.mint(join.owner, mintAmount);
         }
+    }
+
+    function _approvePoolTokenDeposit(
+        AmmData.Context  memory  ctx,
+        uint96                   amount,
+        address                  to
+        )
+        private
+    {
+        require(amount > 0, "INVALID_DEPOSIT_AMOUNT");
+
+        // Check that the deposit in the block matches the expected deposit
+        DepositTransaction.Deposit memory deposit = ctx._block.readDeposit(ctx.txIdx++);
+        ctx.numTransactionsConsumed++;
+
+        AmmData.Token memory poolToken = ctx.tokens[ctx.size];
 
-        S.mint(address(this), poolAmountOut);
-        S.addUserBalance(address(this), join.owner, poolAmountOut);
+        require(
+            deposit.to == to &&
+            deposit.tokenID == poolToken.tokenID &&
+            deposit.amount == amount,
+            "INVALID_TX_DATA"
+        );
+
+        ctx.exchange.deposit{value: 0}(
+            address(this), // from
+            to,
+            poolToken.addr,
+            amount,
+            new bytes(0)
+        );
     }
 
-    function _calculateJoinAmounts(
+    function _removeJoinLock(
         AmmData.State    storage S,
+        AmmData.PoolJoin memory  join
+        )
+        private
+    {
+        AmmData.TokenLock[] storage joinLocks = S.joinLocks[join.owner];
+
+        if (join.nonce == joinLocks.length) {
+            joinLocks.pop();
+        } else {
+            delete joinLocks[join.nonce - 1];
+            if (S.joinLocksStartIdx[join.owner] == join.nonce - 1) {
+                S.joinLocksStartIdx[join.owner]++;
+            }
+        }
```

Comment:
`joinLocksStartIdx` never decreases here (or even in `withdraw` as far as I can tell), so I think it's possible to increase `joinLocksStartIdx`, then the operator processes the latest joins each time. In that case `joinLocks.length == 0` and `joinLocksStartIdx > 0`, and so when the user does another join that join will be missed in `withdraw`.